internal/core/metadata/v2/application/deviceprofile.go

```diff
@@ -86,14 +86,14 @@ func DeleteDeviceProfileByName(name string, ctx context.Context, dic *di.Contain
 		return errors.NewCommonEdgeXWrapper(err)
 	}
 	if len(devices) > 0 {
-		return errors.NewCommonEdgeX(errors.KindContractInvalid, "fail to delete the device profile when associated device exists", nil)
+		return errors.NewCommonEdgeX(errors.KindServiceLocked, "fail to delete the device profile when associated device exists", nil)
```

Comment:
Why `KindServiceLocked` when an associated device exists? Service isn't actually locked, right?

Reply:
Right. Maybe we can create another ErrKind like KindResourceLocked with 423 status code. What do you think?

Reply:
Agreed.

Reply:
@weichou1229 , not sure `423 locked` the correct status for this error?

Reply:
How about `428 Precondition Required` i.e. the precondition is `no associated device`

Reply:
The original design is 423 status code
 https://github.com/edgexfoundry/edgex-go/blob/master/openapi/v2/core-metadata.yaml#L2214

And I found the description on https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses
- **423 Locked**: The resource that is being accessed is locked.
- **428 Precondition Required**: The origin server requires the request to be conditional. This response is intended to prevent the 'lost update' problem, where a client GETs a resource's state, modifies it, and PUTs it back to the server, when meanwhile a third party has modified the state on the server, leading to a conflict.

I thought the description of 423 is more appropriate.

Reply:
IMO, 428 isn't appropriate since the profile ins't locked

```
The 423 (Locked) status code means the source or destination resource of a method is locked. This response SHOULD contain an appropriate precondition or postcondition code, such as 'lock-token-submitted' or 'no-conflicting-lock'.
```

Reply:
@weichou1229 How about:

 _**409 Conflict**
The request could not be completed due to a conflict with the current state of the resource. This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request. The response body SHOULD include enough information for the user to recognize the source of the conflict. Ideally, the response entity would include enough information for the user or user agent to fix the problem; however, that might not be possible and is not required.

Conflicts are most likely to occur in response to a PUT request. For example, if versioning were being used and the entity being PUT included changes to a resource which conflict with those made by an earlier (third-party) request, the server might use the 409 response to indicate that it can't complete the request. In this case, the response entity would likely contain a list of the differences between the two versions in a format defined by the response Content-Type._


User can resolve the conflict by removing the associated Device(s).

Reply:
> The response body SHOULD include enough information for the user to recognize the source of the conflict. Ideally, the response entity would include enough information for the user or user agent to fix the problem

It's great to me. User need enough information to resolve the error.

But it might lead misunderstanding because we have 409 on swagger doc like "Conflict detected. Device profile name must be universally unique."
https://github.com/edgexfoundry/edgex-go/blob/master/openapi/v2/core-metadata.yaml#L2008

What do you think? @cloudxxx8

Reply:
The response message gives it context for the API that was used.

Reply:
Let's use 409 for this kind of scenario and modify the error message properly.

Reply:
@lenny-intel @cloudxxx8  

Currently, we have KindDuplicateName for http.StatusConflict. 
https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/errors/types.go#L196

But the KindDuplicateName doesn't match our error case. We might need to create a new ErrKind with a proper name.
How about **KindConflictState** or **KindResourceOccupied**?

Reply:
`KindStatusConflict` sounds good.

Reply:
yes, please use `KindStatusConflict`

Reply:
Created PR for the new ErrKind https://github.com/edgexfoundry/go-mod-core-contracts/pull/470